src/deploy/functions/checkIam.ts

```diff
@@ -0,0 +1,96 @@
+import { has, last } from "lodash";
+import { bold } from "cli-color";
+
+import { debug } from "../../logger";
+import * as track from "../../track";
+import { getReleaseNames, getFunctionsInfo, getFilterGroups } from "../../functionsDeployHelper";
+import { FirebaseError } from "../../error";
+import { testIamPermissions, testResourceIamPermissions } from "../../gcp/iam";
+
+const PERMISSION = "cloudfunctions.functions.setIamPolicy";
+
+/**
+ * Checks to see if the authenticated account has `iam.serviceAccounts.actAs` permissions
+ * on a specified project (required for functions deployments).
+ * @param projectId The project ID to check.
+ */
+export async function checkServiceAccountIam(projectId: string): Promise<void> {
+  const saEmail = `${projectId}@appspot.gserviceaccount.com`;
+  let passed = false;
+  try {
+    const iamResult = await testResourceIamPermissions(
+      "https://iam.googleapis.com",
+      "v1",
+      `projects/${projectId}/serviceAccounts/${saEmail}`,
+      ["iam.serviceAccounts.actAs"]
+    );
+    passed = iamResult.passed;
+  } catch (err) {
+    debug("[functions] service account IAM check errored, deploy may fail:", err);
+    return;
```

Comment:
If this return is correct, please add a comment indicating why we wouldn't want to re-throw the error.

Reply:
Done.